gateway/src/lib.rs

```diff
@@ -806,30 +813,48 @@ impl Actor {
                 return Err(actor_error!(illegal_state, "owner not match"));
             }
 
-            if rt.message().value_received() < *MIN_CROSS_MSG_GAS {
-                return Err(actor_error!(illegal_state, "not enough gas"));
-            }
+            // collect cross-fee
+            st.collect_cross_fee(&mut value, &*CROSS_MSG_FEE)?;
 
-            let PostBoxItem { cross_msg, .. } = postbox_item;
-            Self::commit_cross_message(rt, st, cross_msg)?;
-            st.remove_from_postbox(rt.store(), postbox_cid)
+            let PostBoxItem { mut cross_msg, .. } = postbox_item;
+            do_burn = Self::commit_cross_message(rt, st, &mut cross_msg)?;
+            st.remove_from_postbox(rt.store(), postbox_cid)?;
+            Ok(cross_msg)
         })?;
 
+        // if bottom_up message being committed.
+        if do_burn {
+            rt.send(
+                *BURNT_FUNDS_ACTOR_ADDR,
+                METHOD_SEND,
+                RawBytes::default(),
+                cross_msg.msg.value.clone(),
+            )?;
+        }
+
+        // send the remainder of the fee to the owner
+        if value > TokenAmount::zero() {
+            rt.send(owner, METHOD_SEND, RawBytes::default(), value.clone())?;
+        }
+
         Ok(RawBytes::default())
     }
 
-    /// Commit the cross message to storage.
+    /// Commit the cross message to storage. It outputs a flag signaling
+    /// if the committed messages was bottom-up and some funds need to be
+    /// burnt.
     ///
     /// NOTE: This function should always be called inside an `rt.transaction`
     fn commit_cross_message<BS, RT>(
         rt: &mut RT,
         st: &mut State,
-        mut cross_msg: CrossMsg,
-    ) -> Result<(), ActorError>
+        mut cross_msg: &mut CrossMsg,
```

Comment:
do we need the starting `mut` here?

Reply:
We do, when we commit a message there may be changes over the nonce and other internal values.